[Observability Onboarding] OTel logs flow

flash1293 opened the pull request:
# Testing instructions

* The k8s flow should work already for logs but won't report any metrics
* The linux/mac flow requires the following stopgap measure until everything is ready
  * Run the provided snippet
  * It will fail with `No such file or directory`
  * Run either `touch otel.yml && mkdir otel_samples && curl https://raw.githubusercontent.com/elastic/elastic-agent/871ad33afc5ac1614f0645e86f2a13c05631aa6d/internal/pkg/otel/samples/darwin/platformlogs_hostmetrics.yml -o otel_samples/platformlogs_hostmetrics.yml` for mac or `touch otel.yml && mkdir otel_samples && curl https://raw.githubusercontent.com/elastic/elastic-agent/871ad33afc5ac1614f0645e86f2a13c05631aa6d/internal/pkg/otel/samples/linux/platformlogs_hostmetrics.yml -o otel_samples/platformlogs_hostmetrics.yml` for linux
  * Re-run the second part of the snippet (the part starting with `rm`)
  * The rest should work as before

Closes https://github.com/elastic/kibana/issues/184433

<img width="931" alt="Screenshot 2024-06-20 at 15 04 19" src="https://github.com/elastic/kibana/assets/1508364/f50392ae-d956-463f-9120-0d01b0fbfa80">


<img width="937" alt="Screenshot 2024-06-20 at 15 04 32" src="https://github.com/elastic/kibana/assets/1508364/a968f615-f173-4210-bb47-6700cc989d5a">



TODOs:
* Real kubernetes snippet
* Do not show on serverless (disabled for testability)
* Remove snapshot version and CDN url

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `run` `docs-build` : Re-trigger the docs validation. (use unformatted text in the comment!)
</p>
</details>

flash1293 commented:
/ci

flash1293 commented:
/ci

flash1293 commented:
/ci

flash1293 commented:
/ci

flash1293 commented:
/ci

flash1293 commented:
/ci

flash1293 commented:
/ci

flash1293 commented:
/ci

strawgate commented:
Everything surrounded by `!` should be replaced before showing the user
Linux:
```
arch=$(if ([[ $(arch) == "arm" || $(arch) == "aarch64" ]]); then echo "arm64"; else echo $(arch); fi)

curl --output elastic-distro-!version!-linux-$arch.tar.gz --url https://!baseurl!/elastic-agent-!version!-linux-$arch.tar.gz --proto '=https' --tlsv1.2 -fOL && tar -xvf elastic-distro-!version!-linux-$arch.tar.gz -C "elastic-distro-!version!-linux-$arch" --strip-components=1 && cd elastic-distro-!version!-linux-$arch 

rm ./otel.yml && cp ./otel_samples/platformlogsandhostmetrics.yml ./otel.yml && sed -i 's#<<ES_ENDPOINT>>#!es_endpoint!#g' ./otel.yml && sed -i 's/<<ES_API_KEY>>/!es_api_key!/g' ./otel.yml
```

macos
```
arch=$(if [[ $(arch) == "arm" ]]; then echo "aarch64"; else echo $(arch); fi)

curl --output elastic-distro-!version!-macos-$arch.tar.gz --url https://!baseurl!/elastic-agent-!version!-macos-$arch.tar.gz --proto '=https' --tlsv1.2 -fOL && tar -xvf elastic-distro-!version!-macos-$arch.tar.gz -C "elastic-distro-!version!-macos-$arch" --strip-components=1 && cd elastic-distro-!version!-macos-$arch 

rm ./otel.yml && cp ./otel_samples/platformlogsandhostmetrics.yml ./otel.yml && sed -i '' 's#<<ES_ENDPOINT>>#!es_endpoint!#g' ./otel.yml && sed -i '' 's/<<ES_API_KEY>>/!es_api_key!/g' ./otel.yml
```

Example rendered one:
```
arch=$(if ([[ $(arch) == "arm" || $(arch) == "aarch64" ]]); then echo "arm64"; else echo $(arch); fi)

curl --output elastic-distro-8.14.0-SNAPSHOT-linux-$arch.tar.gz --url https://snapshots.elastic.co/8.14.0-6b2f3648/downloads/beats/elastic-agent/elastic-agent-8.14.0-SNAPSHOT-linux-$arch.tar.gz --proto '=https' --tlsv1.2 -fOL && tar -xvf elastic-distro-8.14.0-SNAPSHOT-linux-$arch.tar.gz -C "elastic-distro-8.14.0-SNAPSHOT-linux-$arch" --strip-components=1 && cd elastic-distro-8.14.0-SNAPSHOT-linux-$arch 

rm ./otel.yml && cp ./otel_samples/base.yml ./otel.yml && sed -i 's#<<ES_ENDPOINT>>#https://localhost:9200#g' ./otel.yml && sed -i 's/<<ES_API_KEY>>/1234567890/g' ./otel.yml
```

Will work on the Windows one shortly

strawgate commented:
I think `base_url` will be:

```
https://artifacts.elastic.co/downloads/beats/elastic-agent/elastic-agent-!version!-linux-$arch.tar.gz
```

With base_url being `https://artifacts.elastic.co/downloads/beats/elastic-agent/`

strawgate commented:
Running: `./otelcol --config otel.yml`

flash1293 commented:
Updated the PR, preview should be available soon - @strawgate are you sure about the macos snippet? This errors for me for older versions. In the other onboarding flow the download URL for agents on mac is something like `.../elastic-agent-8.8.2-darwin-aarch64.tar.gz` (`darwin` instead of `macos` and `aarch64` instead of `arm64`).

See here for the generation logic: https://github.com/elastic/kibana/blob/1deb6e10197a13798b903ba981700248a92f466e/x-pack/plugins/observability_solution/observability_onboarding/public/assets/standalone_agent_setup.sh#L48

If the scheme of these URLs will change starting from 8.15, we need to make sure all places are adjusted accordingly.

flash1293 commented:
/ci

flash1293 commented:
/ci

strawgate commented:
Sorry yes I think I pasted the wrong snippet, sorry about that. Will try to update today

strawgate commented:
Updated Linux and macOS:

Linux:
```
arch=$(if ([[ $(arch) == "arm" || $(arch) == "aarch64" ]]); then echo "arm64"; else echo $(arch); fi)

curl --output elastic-distro-!version!-linux-$arch.tar.gz --url https://!baseurl!/elastic-agent-!version!-linux-$arch.tar.gz --proto '=https' --tlsv1.2 -fOL && mkdir elastic-distro-!version!-linux-$arch && tar -xvf elastic-distro-!version!-linux-$arch.tar.gz -C "elastic-distro-!version!-linux-$arch" --strip-components=1 && cd elastic-distro-!version!-linux-$arch 

rm ./otel.yml && cp ./otel_samples/platformlogsandhostmetrics.yml ./otel.yml && sed -i 's#<<ES_ENDPOINT>>#!es_endpoint!#g' ./otel.yml && sed -i 's/<<ES_API_KEY>>/!es_api_key!/g' ./otel.yml
```

macOS
```
arch=$(if [[ $(arch) == "arm64" ]]; then echo "aarch64"; else echo $(arch); fi)

curl --output elastic-distro-!version!-darwin-$arch.tar.gz --url https://!baseurl!/elastic-agent-!version!-darwin-$arch.tar.gz --proto '=https' --tlsv1.2 -fOL && mkdir "elastic-distro-!version!-darwin-$arch" && tar -xvf elastic-distro-!version!-darwin-$arch.tar.gz -C "elastic-distro-!version!-darwin-$arch" --strip-components=1 && cd elastic-distro-!version!-darwin-$arch 

rm ./otel.yml && cp ./otel_samples/platformlogsandhostmetrics.yml ./otel.yml && sed -i '' 's#<<ES_ENDPOINT>>#!es_endpoint!#g' ./otel.yml && sed -i '' 's/<<ES_API_KEY>>/!es_api_key!/g' ./otel.yml
```

flash1293 commented:
/ci

elasticmachine commented:
Pinging @elastic/obs-ux-logs-team (Team:obs-ux-logs)

flash1293 commented:
/ci

strawgate commented:
Windows + very minor change to linux and macos (switch `platformlogsandhostmetrics.yml` to `platformlogs_hostmetrics.yml`).

Windows:
```
$ProgressPreference = 'SilentlyContinue'

Invoke-WebRequest -OutFile elastic-distro-!version!-windows-x86_64.zip -Uri https://!baseurl!/elastic-agent-!version!-windows-x86_64.zip
Expand-Archive .\elastic-distro-!version!-windows-x86_64.zip -DestinationPath .; Move-Item .\elastic-agent-!version!-windows-x86_64 .\elastic-distro-!version!-windows-x86_64; Set-Location elastic-distro-!version!-windows-x86_64

remove-item ./otel.yml; copy-item ./otel_samples/hostmetrics.yml ./otel.yml; ((Get-Content ./otel.yml) -replace '<<ES_ENDPOINT>>', '!es_endpoint!') -replace '<<ES_API_KEY>>', '!es_api_key!' | Set-Content ./otel.yml

.\otelcol.exe --config .\otel.yml
```

Updated Linux and macOS:

Linux:
```
arch=$(if ([[ $(arch) == "arm" || $(arch) == "aarch64" ]]); then echo "arm64"; else echo $(arch); fi)

curl --output elastic-distro-!version!-linux-$arch.tar.gz --url https://!baseurl!/elastic-agent-!version!-linux-$arch.tar.gz --proto '=https' --tlsv1.2 -fOL && mkdir elastic-distro-!version!-linux-$arch && tar -xvf elastic-distro-!version!-linux-$arch.tar.gz -C "elastic-distro-!version!-linux-$arch" --strip-components=1 && cd elastic-distro-!version!-linux-$arch 

rm ./otel.yml && cp ./otel_samples/platformlogs_hostmetrics.yml ./otel.yml && sed -i 's#<<ES_ENDPOINT>>#!es_endpoint!#g' ./otel.yml && sed -i 's/<<ES_API_KEY>>/!es_api_key!/g' ./otel.yml
```

macOS
```
arch=$(if [[ $(arch) == "arm64" ]]; then echo "aarch64"; else echo $(arch); fi)

curl --output elastic-distro-!version!-darwin-$arch.tar.gz --url https://!baseurl!/elastic-agent-!version!-darwin-$arch.tar.gz --proto '=https' --tlsv1.2 -fOL && mkdir "elastic-distro-!version!-darwin-$arch" && tar -xvf elastic-distro-!version!-darwin-$arch.tar.gz -C "elastic-distro-!version!-darwin-$arch" --strip-components=1 && cd elastic-distro-!version!-darwin-$arch 

rm ./otel.yml && cp ./otel_samples/platformlogs_hostmetrics.yml ./otel.yml && sed -i '' 's#<<ES_ENDPOINT>>#!es_endpoint!#g' ./otel.yml && sed -i '' 's/<<ES_API_KEY>>/!es_api_key!/g' ./otel.yml
```

It might be worth noting that these instructions do not install the collector, just download and run.

flash1293 commented:
Thanks @strawgate , added that. I changed the wording slightly to make it clear it's not really installing the collector and how the second command is starting it

strawgate commented:
I think installing the collector is just a single command away, ill do some testing on this and report back perhaps we can offer both

thomheymann reviewed (commented):
Not reviewed the code yet but couple comments regarding the UX below:

<img width="1151" alt="Screenshot 2024-06-17 at 11 33 17" src="https://github.com/elastic/kibana/assets/190132/e81e4eae-94b8-4a5d-9297-e3ec93b2cded">

flash1293 commented:
Thanks for the comments, very helpful @thomheymann !

@mlunadia 

Showing the K8s tab by default makes sense to me, wdyt?

sileschristian commented:
@thomheymann Thank you for your feedback. I think that copy above the CTA is just a mistake. This is the latest proposal we made -> [Design](https://www.figma.com/design/CPhMyRNOgo0wsEiaIMZJ14/Onboarding-Quick-Starts?node-id=1185-15332&t=7o3ANKrVf4X9fGdP-4)

flash1293 commented:
@sileschristian @thomheymann 
* Made k8s the default
* Small change - first step is now called `Download and run the collector` (as we don't install in case of linux/windows/mac)
* Removed the copy for hosts, but for kubernetes I think we somehow need to represent what the user needs to do with the manifest. I added this:
<img width="1111" alt="Screenshot 2024-06-18 at 17 10 53" src="https://github.com/elastic/kibana/assets/1508364/2254dcad-ffc0-47c3-b28d-8149dbf7a176">


The copy for the second code block from the design is wrong - "Run the following command to make sure the collector is running properly." is not what's happening here, this second snippet is to actually start the collector. The idea to split it out is that the user might run this multiple times, but they won't for the first part. No strong opinion on making it a separate step, what do you think @sileschristian ?

mlunadia commented:
>@mlunadia
Showing the K8s tab by default makes sense to me, wdyt?

100% the order should be as the [priority given above](https://github.com/elastic/observability-dev/issues/3217#issuecomment-2129933019)

mlunadia commented:
>kubernetes I think we somehow need to represent what the user needs to do with the manifest. 

@flash1293 let's follow the same pattern/copy we have for Elastic Agent
<img width="796" alt="image" src="https://github.com/elastic/kibana/assets/39376769/9eff76a9-7a67-4482-9352-211fb5bddf80">

flash1293 commented:
Makes sense to me, adjusted here:

<img width="1149" alt="Screenshot 2024-06-18 at 17 50 59" src="https://github.com/elastic/kibana/assets/1508364/56206a69-2498-4818-8c60-b33f29d96cf9">

mlunadia commented:
<img width="182" alt="image" src="https://github.com/elastic/kibana/assets/39376769/6a104601-15fe-4bd8-90ea-7e3e242a3e67">

@flash1293 I'd replace this with 'Download manifest' as this is not just the collector config.

sileschristian commented:
> No strong opinion on making it a separate step, what do you think @sileschristian ?

If it's a command to make sure the collector is running, this feels a bit optional to me. We can break it on it's own step: **2. To ensure the collector is running properly, run the following command.** 

@flash1293 Can we avoid using the little copy icon on the right? It gets a bit lost. I'd suggest adding always a CTA under the command.  `

sileschristian commented:
Thanks for sharing @flash1293. Based on your feedback I'm proposing this:

1. I think we should keep the minimal steps possible. In order to not have one step for selecting the platform and one to copy command/download manifest I'd rather add a simple line of text above the code that changes depending on the platform selection.
2. Since there is no main cta because all steps are active and we can't control how the user clicks on them. All buttons have the same weight (e.g. copy or download and then copy and copy). I'd suggest to use an standard button ([read more here](https://eui.elastic.co/#/navigation/button/guidelines:~:text=This%20button%20type%20works%20well%20for%20multiple%20actions%20of%20equal%20weight.))

3. Last step can keep the same for now.

![image](https://github.com/elastic/kibana/assets/127108526/23cd986b-a17e-49f4-ba35-eeabd5b27c42)
![image](https://github.com/elastic/kibana/assets/127108526/f9162293-8c85-487e-a1c9-ff9935d82a72)

[Figma link](https://www.figma.com/design/CPhMyRNOgo0wsEiaIMZJ14/Onboarding-Quick-Starts?node-id=1810-8995&t=vfnTZ4azG6F8STlX-4)


What do you think?

flash1293 commented:
Thanks @sileschristian , updated by your comments:

<img width="931" alt="Screenshot 2024-06-20 at 15 04 19" src="https://github.com/elastic/kibana/assets/1508364/f50392ae-d956-463f-9120-0d01b0fbfa80">


<img width="937" alt="Screenshot 2024-06-20 at 15 04 32" src="https://github.com/elastic/kibana/assets/1508364/a968f615-f173-4210-bb47-6700cc989d5a">


I also removed windows as we don't support it right now and use the stack version on stateful instead of the latest available agent version.

sileschristian commented:
Thank you @flash1293. Looks good to me. When we have the [header for the K8s](https://github.com/elastic/kibana/issues/186397) flow we could also improve the navigation here and apply the same header.

thomheymann commented on x-pack/plugins/observability_solution/observability_onboarding/public/application/observability_onboarding_flow.tsx:
```diff
@@ -60,6 +61,10 @@ export function ObservabilityOnboardingFlow() {
               <BackButton />
               <CustomLogsPanel />
             </Route>
+            <Route path="/otelLogs">
```
nit: I appreciate the old Quickstart flows used camelCasing for URLs but Kibana and ES use underscores/hyphens to separate words in URLs so would be good to follow that for consistency

thomheymann commented on x-pack/plugins/observability_solution/observability_onboarding/public/assets/dashboard_illustration.svg:
The drop shadow should not be part of the SVG since it makes it impossible to size and position the actual image correctly otherwise. 

You can use the existing waterfall SVG in the `assets/` folder and then render it with `<EuiImage hasShadow />` to get the correct drop shadow.

thomheymann commented on x-pack/plugins/observability_solution/observability_onboarding/public/application/onboarding_flow_form/use_custom_cards_for_category.ts:
```diff
@@ -168,7 +191,8 @@ export function useCustomCardsForCategory(
           integration: '',
         },
-        toFeaturedCard('nginx'),
+        // TODO enable before merge: isServerless ? toFeaturedCard('nginx') : otelCard,
+        otelCard,
```
TODO before merge :)

strawgate commented:
@flash1293 for this comment https://github.com/elastic/elastic-agent/pull/5002 the snippets won't have `SNAPSHOT` in them for release right?

flash1293 commented:
@strawgate right, that's just for testing right now, I will remove it before merge.

thomheymann reviewed (changes requested):
Overall looks good but found a few issues:

The generated command should not contain any unescaped line breaks. When executing the K8 command it currently returns the following error message because of the line break:

```
zsh: command not found: --from-literal=es_api_key=xxxxxxxxx
```

The secret store seems to clash with the existing K8 flow - I'm getting the following error when running the command after having completed the K8 flow that Mykola has been working on:

```
error: failed to create secret secrets "elastic-secret" already exists
```

You've added a new card to the suggestions panels without removing one of the existing one's so the grid now breaks awkwardly. According to the designs the grid should be limited to 6 cards:

<img width="1198" alt="Screenshot 2024-07-02 at 10 07 52" src="https://github.com/elastic/kibana/assets/190132/1d9689ff-6630-405b-8fcc-7bbc7652fb5b">

flash1293 commented:
Thanks for the review @thomheymann - I addressed your points:
* Fixed the secret creation command
* Making sure the secret name doesn't collide with the other onboarding flow - entirely possible people will test-drive both, good catch!
* Removed the duplicated tile

thomheymann reviewed (approved):
LGTM! ✨

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/219165)
* Commit: f3ac5ef91611aeff228fb01049099dae88b81a80
* Kibana Serverless Image: `docker.elastic.co/kibana-ci/kibana-serverless:pr-183732-f3ac5ef91611`

### Failed CI Steps


<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/219165","number":219165,"commit":"f3ac5ef91611aeff228fb01049099dae88b81a80"}],"number":219165}
buildkite-pr-comment-->
